clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -287,6 +287,7 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
             this.sender = new Sender(client,
                     this.metadata,
                     this.accumulator,
+                    config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION) == 1,
```

Comment:
Sounds good to me as well. It looks less hacky. Originally I was thinking about a new configuration like ```max.batch.per.partition```. I was worrying about confusing user if we reuse ```max.in.flight.request.per.connection``` and it seems not necessary to enforce max in flight batch when ```max.in.flight.request.per.connection``` is greater than 1.